src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs

```diff
@@ -76,296 +67,106 @@ public override async Task Dispatch(PeerInfo bestPeer, BlocksRequest? blocksRequ
             }
 
-            if (bestPeer is null)
-            {
-                string message = $"Not expecting best peer to be null inside the {nameof(BlockDownloader)}";
-                if (_logger.IsError) _logger.Error(message);
-                throw new ArgumentNullException(message);
-            }
-
-            DownloaderOptions options = blocksRequest.Options;
-            bool downloadReceipts = (options & DownloaderOptions.WithReceipts) == DownloaderOptions.WithReceipts;
-            bool shouldProcess = (options & DownloaderOptions.Process) == DownloaderOptions.Process;
-            bool shouldMoveToMain = (options & DownloaderOptions.MoveToMain) == DownloaderOptions.MoveToMain;
-
-            int blocksSynced = 0;
-            long currentNumber = _blockTree.BestKnownNumber;
-            if (_logger.IsDebug)
-                _logger.Debug(
-                    $"MergeBlockDownloader GetCurrentNumber: currentNumber {currentNumber}, beaconPivotExists: {_beaconPivot.BeaconPivotExists()}, BestSuggestedBody: {_blockTree.BestSuggestedBody?.Number}, BestKnownNumber: {_blockTree.BestKnownNumber}, BestPeer: {bestPeer}, BestKnownBeaconNumber {_blockTree.BestKnownBeaconNumber}");
+            return await base.DownloadBlocks(bestPeer, blocksRequest, cancellation);
+        }
 
-            bool HasMoreToSync(out BlockHeader[]? headers, out int headersToRequest)
+        public override async Task<long> DownloadHeaders(PeerInfo? bestPeer, BlocksRequest blocksRequest,
+            CancellationToken cancellation)
+        {
+            // Note: Redundant with Dispatch, but test uses it.
+            if (_beaconPivot.BeaconPivotExists() == false && _poSSwitcher.HasEverReachedTerminalBlock() == false)
             {
                 if (_logger.IsDebug)
-                    _logger.Debug($"Continue full sync with {bestPeer} (our best {_blockTree.BestKnownNumber})");
-
-                headersToRequest = Math.Min(_syncBatchSize.Current, bestPeer.MaxHeadersPerRequest());
-                if (_logger.IsTrace)
-                    _logger.Trace(
-                        $"Full sync request {currentNumber}+{headersToRequest} to peer {bestPeer} with {bestPeer.HeadNumber} blocks. Got {currentNumber} and asking for {headersToRequest} more.");
+                    _logger.Debug("Using pre merge block downloader");
+                return await _preMergeBlockDownloader.DownloadHeaders(bestPeer, blocksRequest, cancellation);
+            }
 
-                headers = _chainLevelHelper.GetNextHeaders(headersToRequest, bestPeer.HeadNumber, blocksRequest.NumberOfLatestBlocksToBeIgnored ?? 0);
-                if (headers is null || headers.Length <= 1)
-                {
-                    if (_logger.IsTrace)
-                        _logger.Trace("Chain level helper got no headers suggestion");
-                    return false;
-                }
+            return await base.DownloadHeaders(bestPeer, blocksRequest, cancellation);
+        }
```

Comment:
We could potentially move all this structure with passing a delegate, but might not be worth it.